### PR TITLE
feat(frontend): table flows

### LINKED ISSUE #1007: 🚀 [Feature] Frontend: Create own table with settings
## 🚀 Feature
Who: users
What: Create own table, with chosen name and visibility / access rights
Why: I might want to give a hint about what my table is about. I might want to be found by anyone interested, or have a private talk with friends.

The name should be used in deep links as well (#1581 ).

### Design Templates
https://www.figma.com/design/nkCyyDFx7fGeBnguwpardz/Dreammall-pr%C3%A4si?node-id=3426-47324&t=hfsalBftC4qnbKwX-4
https://www.figma.com/design/nkCyyDFx7fGeBnguwpardz/Dreammall-pr%C3%A4si?node-id=3438-33901&t=hfsalBftC4qnbKwX-4

Image
Image
Image
Image
Image

## Todos
- [ ] Let desktop button float at bottom of page
- [ ] Unify action list of desktop and mobile
- [ ] New action list to create own table (or open it, should it already exist)
- [ ] Implement dialog
- [ ] ...?

Backend
------

- [x] #1448
- [ ] Get list of users I can add to the table?

### LINKED ISSUE #1593: 🚀 [Feature] Frontend: Change own table settings
## 🚀 Feature

As a user, I want to change the settings of my existing own table because I want to chose a new topic and hence name, or might want to change who can join my table.

Screenshot 2024-07-31 at 12 24 01
Screenshot 2024-07-31 at 12 24 05

Related to https://github.com/dreammall-earth/dreammall.earth/issues/1361

## Todo
- [ ] Clarify dialogs @Hardy-Hart-am-Start 
- [x] Possibility to have different actions for DreamMall Button on different pages

### PR DESCRIPTION
**Pull request by sebastian2357**
## 🍰 Pullrequest

- This PR shows the current state of table flow development.
-  It is NOT ready for review! But feedback is more than welcome :)

### Issues

- fixes #1007
- fixes #1593 

### Todo

- [ ] GraphQL
- [ ] Dialoge for an active session
- [ ] Clean Up

### CODE REVIEW

File: frontend/src/panels/dreammall/TableSetup.vue [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+    @back="onBack"
+    @close="onClose"
+  />
+  <component
+    :is="steps[currentStep].component"
+    v-if="steps && currentStep < steps.length"
+    :my-table-settings="tableSettings"
+    @next="onNext"
+    @submit="onSubmit"
+    @table-name:updated="updateTableName"
```
Comment (Bettelstab) [inline frontend/src/panels/dreammall/TableSetup.vue:16]: I wonder if `v-model=tableSettings` would do the trick.

File: frontend/src/panels/dreammall/TableSetup.vue [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+const emit = defineEmits<{
+  (e: 'close'): void
+}>()
+
+const updateHistory = (step: number) => {
+  // if (window?.location?.href == null) return
+  // const url = new URL(window.location.href)
+  // url.searchParams.set('step', step.toString())
+  // todo: find other solution. This causes conflicts with Vike
+  // window.history.pushState({ step }, '', url.toString())
```
Comment (Bettelstab) [inline frontend/src/panels/dreammall/TableSetup.vue:99]: I would like to try https://vike.dev/navigate#history-pushstate, but can be after we merge this PR.

File: frontend/src/panels/dreammall/TableSetup.vue [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+    updateHistory(currentStep.value)
+  } else {
+    emit('close')
+  }
+}
+
+const findStepById = (id: string): number => steps.findIndex((step) => step.id === id)
+
+const onNext = transitToNext
+const onBack = transitToPrevious
```
Comment (Bettelstab) [inline frontend/src/panels/dreammall/TableSetup.vue:136]: Why not `@next="transitToNext"`  ?

File: frontend/src/panels/dreammall/TableSetupStepA.vue [hunk 4]
```diff
@@ -0,0 +1,17 @@
+<template>
+  <SimpleButton label="Tisch erstellen" icon="mdi-video" class="mb-5" @click="onClick" />
+</template>
+
+<script setup lang="ts">
+import SimpleButton from '#components/buttons/SimpleButton.vue'
+import { TableSetupEmits, TableSetupProps } from '#src/panels/dreammall/TableSetupProps'
+
+const props = defineProps<TableSetupProps>()
```
Comment (Bettelstab) [inline frontend/src/panels/dreammall/TableSetupStepA.vue:9]:
```suggestion
defineProps<TableSetupProps>()
```

**Review by Bettelstab — CHANGES_REQUESTED** [review]
I think there should be some adjustments code-wise.
But the most important, I have problems testing. User search does not work, and after I added my table, I get 500 from the frontend server.
Maybe my merge was bad, we need to check this in any case.

File: frontend/src/panels/dreammall/TableSetupProps.ts [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+
+export type TableSetupProps = {
+  submitText: string
+  myTableSettings: MyTableSettings
+}
+
+export type TableSetupEmits = {
+  (e: 'next'): void
+  (e: 'submit'): void
+  (e: 'custom', stepId: string): void
```
Comment (Bettelstab) [inline frontend/src/panels/dreammall/TableSetupProps.ts:11]: I think 'goTo' would be a better name.

File: frontend/src/panels/dreammall/TableSetupStepB.vue [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+  },
+})
+
+const onNext = () => {
+  emit('next')
+}
+</script>
+
+<style lang="scss">
+.flat-text-field .v-field--variant-solo-filled {
```
[the remark this answers is not in the document]
Comment (Bettelstab) [reply frontend/src/panels/dreammall/TableSetupStepB.vue:51]: Please change for clarity

File: frontend/src/stores/userStore.ts [hunk 7]
```diff
@@ -87,7 +98,15 @@ export const useUserStore = defineStore(
 
     const getCurrentUser = computed(() => currentUser.value)
 
-    const getMyTable = computed(() => currentUser.value?.table)
+    const getMyTable = computed(
+      () =>
+        currentUser.value?.table || {
```
Comment (Bettelstab) [inline frontend/src/stores/userStore.ts:103]: Can't it be null?

File: frontend/src/stores/userStore.ts [hunk 8]
```diff
@@ -184,6 +195,17 @@ export const useUserStore = defineStore(
       }
     }
 
+    const searchUsers = (name: string) => {
```
[the remark this answers is not in the document]
Comment (Bettelstab) [reply frontend/src/stores/userStore.ts:206]: still here?!

**Review by Bettelstab — CHANGES_REQUESTED** [review]
For me it's fine to merge all the new files as they are and to improve afterwards - under the given circumstances. But we should be careful with everything affecting existing code / being in existing files, because cleaning that afterwards is more difficult.

File: frontend/src/panels/dreammall/TableSetupProps.ts (hunk 5 above)
Comment (Bettelstab) [reply frontend/src/panels/dreammall/TableSetupProps.ts:11]: (optional for now)

File: frontend/src/stores/userStore.ts (hunk 7 above)
Comment (Bettelstab) [reply frontend/src/stores/userStore.ts:103]: must have to clarify

File: frontend/src/panels/dreammall/TableSetupStepC.vue [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+  }
+  invitedUserIdsModel.value = currentUsers
+}
+
+const onNext = () => {
+  emit('next')
+}
+</script>
+
+<style lang="scss">
```
[the remark this answers is not in the document]
Comment (Bettelstab) [reply frontend/src/panels/dreammall/TableSetupStepC.vue:94]: This should also be done before merging please

File: backend/src/auth/authChecker.spec.ts [hunk 10]
```diff
@@ -125,7 +125,7 @@ describe('authChecker', () => {
             introduction: null,
             availability: null,
             // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
-            meetingId: expect.any(Number),
+            meetingId: null,
```
[the remark this answers is not in the document]
Comment (sebastian2357) [reply backend/src/auth/authChecker.spec.ts]: it is a merge from moritz branch

File: frontend/src/layouts/DefaultLayout.vue [hunk 11]
```diff
@@ -53,31 +53,7 @@
         :class="[isButtonListVisible ? 'menu-triangle--turned' : '']"
         :src="Triangle"
       />
-      <MainButton
```
[the remark this answers is not in the document]
Comment (sebastian2357) [reply frontend/src/layouts/DefaultLayout.vue:56]: it is already removed

File: frontend/src/panels/dreammall/TableSetupStepC.vue [hunk 12]
```diff
@@ -0,0 +1,104 @@
+<template>
+  <div class="flat-text-field d-flex flex-column align-center pa-4 w-100">
+    <v-text-field
+      v-model="userSearch"
+      rounded
+      class="elevation-0 w-100"
+      content-class="elevation-0"
+      label="Suche"
```
[the remark this answers is not in the document]
Comment (sebastian2357) [reply frontend/src/panels/dreammall/TableSetupStepC.vue:9]: done

File: frontend/src/panels/PanelFlow.vue [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+
+<!--    return {-->
+<!--      currentStep,-->
+<!--      componentProps,-->
+<!--      goNext,-->
+<!--      goBack,-->
+<!--    }-->
+<!--  },-->
+<!--})-->
+<!--</script>-->
```
[the remark this answers is not in the document]
Comment (sebastian2357) [reply frontend/src/panels/PanelFlow.vue:125]: it prepared for the next refacotring to become a "base compoment". I commented it out, that it does not any effect to the remaining project.

File: frontend/src/panels/PanelHeader.vue [hunk 14]
```diff
@@ -0,0 +1,64 @@
+<template>
+  <div class="step-header">
+    <div class="step-header__left">
+      <v-btn v-if="isBackButtonVisible" icon variant="text" size="small" @click="emitBack">
+        <v-icon>mdi-arrow-left</v-icon>
```
[the remark this answers is not in the document]
Comment (sebastian2357) [reply frontend/src/panels/PanelHeader.vue]: done

File: frontend/src/panels/dreammall/TableSetup.vue (hunk 1 above)
Comment (sebastian2357) [reply frontend/src/panels/dreammall/TableSetup.vue:17]: done

File: frontend/src/panels/dreammall/TableSetup.vue (hunk 2 above)
Comment (sebastian2357) [reply frontend/src/panels/dreammall/TableSetup.vue]: yes

File: frontend/src/panels/dreammall/TableSetup.vue [hunk 15]
```diff
@@ -0,0 +1,203 @@
...
+
+const onSubmit = async () => {
+  try {
+    const name = tableSettings.value.name
+    const isPublic = tableSettings.value.isPublic
+    const users = tableSettings.value.users
```
[the remark this answers is not in the document]
Comment (sebastian2357) [reply frontend/src/panels/dreammall/TableSetup.vue]: done

File: frontend/src/panels/dreammall/TableSetup.vue [hunk 16]
```diff
@@ -0,0 +1,203 @@
...
+    isNaN(initialStep) || initialStep < 0 || initialStep >= steps.length ? 0 : initialStep
+  updateHistory(currentStep.value)
+})
+
+onUnmounted(() => {
+  window.removeEventListener('popstate', handlePopState)
+})
+
+// todo: add createTable
+// todo: hasOwnTable -> getOwnTable
```
[the remark this answers is not in the document]
Comment (sebastian2357) [reply frontend/src/panels/dreammall/TableSetup.vue]: removed

File: frontend/src/panels/dreammall/TableSetupStepA.vue (hunk 4 above)
Comment (sebastian2357) [reply frontend/src/panels/dreammall/TableSetupStepA.vue]: done

File: frontend/src/panels/dreammall/TableSetupStepB.vue (hunk 6 above)
[the remark this answers is not in the document]
Comment (sebastian2357) [reply frontend/src/panels/dreammall/TableSetupStepB.vue:51]: done

File: frontend/src/panels/dreammall/TableSetupStepC.vue (hunk 9 above)
[the remark this answers is not in the document]
Comment (sebastian2357) [reply frontend/src/panels/dreammall/TableSetupStepC.vue:94]: done

File: frontend/src/panels/PanelHeader.vue [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+      </v-btn>
+    </div>
+    <div class="step-header__middle">
+      <h2 class="text-h6">{{ title }}</h2>
+    </div>
+    <div class="step-header__right">
+      <v-btn v-if="isCloseButtonVisible" icon variant="text" size="small" @click="emitClose">
+        <v-icon>mdi-close</v-icon>
```
[the remark this answers is not in the document]
Comment (sebastian2357) [reply frontend/src/panels/PanelHeader.vue]: done

File: frontend/src/panels/dreammall/components/UserInvitationItem.vue [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+
+    <template #append>
+      <v-btn
+        elevation="0"
+        :color="user.invited ? 'success' : 'secondary'"
+        class="invitation-button text-caption py-2 px-3 rounded-pill"
+        size="small"
+        @click="toggleInvitation"
+      >
+        {{ user.invited ? 'gesendet' : 'einladen' }}
```
[the remark this answers is not in the document]
Comment (sebastian2357) [reply frontend/src/panels/dreammall/components/UserInvitationItem.vue]: done

**Review by Bettelstab — APPROVED** [review]

**Pull request merged**